src/rt/array/append.d

```diff
@@ -0,0 +1,290 @@
+ref T[] _d_arrayappendT(T)(return scope ref T[] x, scope T[] y) @trusted
+{
+    // These are needed to fake the safety level to `@safe pure nothrow`.
+    // Without this code will break as these hooks where previously exported
+    // as having this safety level.
+    ref T[] impl(return scope ref T[] x, scope T[] y) @trusted
+    {
+        import core.stdc.string : memcpy;
+        import core.internal.traits : Unqual;
+
+        auto length = x.length;
+        auto sizeelem = T.sizeof;
+
+        _d_arrayappendcTX(x, y.length);
+        if (y.length)
+            memcpy(cast(Unqual!T *)&x[length], cast(Unqual!T *)&y[0], y.length * sizeelem);
+
+        // do postblit
+        __doPostblit(cast(Unqual!T[])x[length .. length + y.length]);
+        return x;
+    }
+
+    alias func = ref T[] delegate(return scope ref T[] x, scope T[] y) @trusted pure nothrow;
+    auto f = cast(func)&impl;
+    return f(x, y);
+}
+
+
+/**
+ * TraceGC wrapper around $(REF _d_arrayappendT, rt,array,append).
+ * Bugs:
+ *   The safety level of this function is faked. It shows itself as `@trusted pure nothrow` to not break existing code.
+ */
+ref T[] _d_arrayappendTTrace(T)(string file, int line, string funcname, return scope ref T[] x, scope T[] y) @trusted
+{
+    ref T[] impl(string file, int line, string funcname, return scope ref T[] x, scope T[] y) @trusted
+    {
+        import rt.profilegc : accumulate;
+        import core.memory : GC;
+
+        static if (is(T == shared))
+            string name = shared(T[]).stringof;
+        else
+            string name = T[].stringof;
+
+        // FIXME: use rt.tracegc.accumulator when it is accessable in the future.
+        version (tracegc)
+        {
+            import core.stdc.stdio;
+
+            printf("%s file = '%.*s' line = %d function = '%.*s' type = %.*s\n",
+                __FUNCTION__.ptr,
+                file.length, file.ptr,
+                line,
+                funcname.length, funcname.ptr,
+                name.length, name.ptr
+            );
+        }
+
+        ulong currentlyAllocated = GC.stats().allocatedInCurrentThread;
+
+        scope(exit)
+        {
+            ulong size = GC.stats().allocatedInCurrentThread - currentlyAllocated;
+            if (size > 0)
+                accumulate(file, line, funcname, name, size);
+        }
+        return _d_arrayappendT(x, y);
+    }
+
+    alias func = ref T[] delegate(string file, int line, string funcname, return scope ref T[] x, scope T[] y) @trusted pure nothrow;
+    auto f = cast(func)&impl;
+    return f(file, line, funcname, x, y);
+}
+
+// This function cannot be imported because then rt.lifetime, and its dependencies needs to be included in the druntime
+// release. It also cannot be core.internal.externDFunc'd because that causes recursive template instantiation.
+
+/// See $(REF _d_arrayappendcTX, rt,lifetime,_d_arrayappendcTX)
+private extern (C) byte[] _d_arrayappendcTX(const TypeInfo ti, ref byte[] px, size_t n);
```

Comment:
I think I commented about this before, but I can't find my comment now.  I think this can be imported instead of being declared as an `extern`

Reply:
I don't think it can because it will be instantiated inside the template and rt.lifetime not part of a dmd installation.

Reply:
Ok, it looks like we'll eventually need to change the DMD installation to make all this templating of druntime work, but that is out of scope of the task at hand.